docs/whatsnew/7.0.rst

```diff
@@ -64,6 +64,13 @@ convention conforms to the `ECSV specification
 existing ECSV readers. Previously the ``meta`` attribute could be written as an ordinary
 YAML map, which is not guaranteed to preserve the order of the keys.
 
+
+Typing in astropy
+=================
+
+The ``astropy.stats`` module is now fully typed. This is the first subpackage for
+which this the case, and we encourage compatibility tests.
```

Comment:
What do you mean by "compatibility tests"? Downstream packages cannot use these annotations for type checking unless we add [a `py.typed` marker file](https://peps.python.org/pep-0561/#packaging-type-information) or [publish a separate package with type stubs](https://peps.python.org/pep-0561/#stub-only-packages).

Reply:
@mhvk suggested this, so I'm not sure exactly what he had in mind. But I don't expect it would be an issue for me to remove it if you'd prefer.

Reply:
I mostly meant that when people type things they presumably would get an error if they pass in the wrong thing. But if that really should have worked, they can let us know, so we fix the typing on our side. Perhaps this can be made more explicit?

Separately, I think numpy does some internal typing tests - maybe we should too? (Not for this PR, though!)

Reply:
p.s. Given @eerovaher's points, my suggestion is perhaps simply impossible, so also fine to remove.

Reply:
I don't think we can ask our users for feedback on type-**checking** until we have some kind of type checker running on our side. As @eerovaher pointed out, we don't currently ship a `py.typed` marker file, which means that our type hints are still invisible to downstream type checkers (not to IDEs !); and they should be, as long as we don't type check for ourselves.

Reply:
Ok I removed the clause 'and we encourage compatibility tests' in [30179c5](https://github.com/astropy/astropy/pull/16562/commits/30179c56e611b3918a273621cfbae544224dde90). Everyone makes good points, worth coming back to regarding potential future type checking (#15170).

Reply:
@eerovaher , are you happy with the changes or is there more? Thanks!

Reply:
I wouldn't add any typing related change log or What's New entries until we are type checking.

Reply:
I see your point, though I think the other user benefits of annotating the code (tab-completion, aiding user understanding, indicating that the codebase is moving toward being annotated e.g. for type checking) are worth noting in a change log entry (and at the scale of a whole subpackage in a What's New entry). Especially as type checking potentially won't happen for quite a while. In a conversation [above](https://github.com/astropy/astropy/pull/16562#discussion_r1639466720), Tom thought the change log entry was warranted, Marten suggested the What's New, and Nathaniel was in favor. I understand these entries haven't been used for previous PRs adding annotations, but having a whole subpackage annotated could be a good time to start.

Reply:
FWIW I do think this is worth mentioning this in the What's New and changelog! I think it's important to communicate to the dev-savvy users that read these things that we are moving (slowly, but still) in the direction of more typing throughout the core package!

Reply:
We used to have a page in the documentation about the API stability in different sub-package (until a932250a1894f29e383647058c44bca8e06de201). Maybe we should have an analogous page that would describe to what extent are the different sub-packages typed?

Reply:
> Maybe we should have an analogous page that would describe to what extent are the different sub-packages typed?

Might actually be more useful than having that info in change logs. 🤔

Reply:
That sounds good.  https://mypy.readthedocs.io/en/stable/command_line.html#report-generation might be good for autodocumenting per-module typing coverage.

Reply:
In [84b0098](https://github.com/astropy/astropy/pull/16562/commits/84b00983ad23e57cfa95ef109e4fa099f1c613bc) I removed the changelog entry. Keeping the What's New entry.

Documenting typing coverage is a good idea, and is beyond the scope of this PR. 

I see this as now ready to merge.

Reply:
I think having a docs page to track type support is good, but should be a follow-up and shouldn't hold up this PR -- could someone make an issue to track that?

Reply:
#16660